File: 6_kyu/Ranking_NBA_teams.go
package kata

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// matchData holds the string encoded results of a match.
type matchData struct {
	team   [2]string
	points [2]string
}

// teamData holds the points scored and conceded in match by a team.
type teamData struct {
	pointsScored, pointsConceded int
}

// The points are given as decimal floating point numbers.
// The points must be the only decimal numbers in the result string.
var patternPoints = regexp.MustCompile(`\b\d+\.?\d*\b`)

// splitMatchData splits the string of one game result into its parts.
// Extracts only the strings.
// Allows floating point numbers as points.
func splitMatchData(game string) (parsed matchData, err error) {
	// The points structed the result string
	indexPoints := patternPoints.FindAllStringIndex(game, -1)
	if indexPoints == nil || len(indexPoints) != 2 {
		err = errors.New("not 2 numbers")
		return
	}
	for id, index := range indexPoints {
		parsed.points[id] = game[index[0]:index[1]]
	}
	parsed.team[0] = strings.TrimSpace(game[0:indexPoints[0][0]])
	parsed.team[1] = strings.TrimSpace(game[indexPoints[0][1]:indexPoints[1][0]])
	return
}

// parsePoints parsed an decimal number.
// Returns float number error if the decimal number is invalid.
func parsePoints(str string) (n int, err error) {
	n, err = strconv.Atoi(str)
	if err != nil {
		err = errors.New("float number")
	}
	return
}

// parseTeamData parses the points of a team.
// The points must be given as integers.
func parseTeamData(str matchData, teamId int) (game teamData, err error) {
	game.pointsScored, err = parsePoints(str.points[teamId])
	if err == nil {
		game.pointsConceded, err = parsePoints(str.points[1-teamId])
	}
	return
}

func NbaCup(resultSheet, toFind string) string {
	if toFind == "" {
		// Special case: empty team must result in an empty string.
		return ""
	}
	// variables to collect game results
	numberMatchesWon := 0
	numberMatchesDrawn := 0
	numberMatchesLost := 0
	totalPointsScored := 0
	totalPointsConceded := 0
	rank := 0
	// Iterate over all games.
	for _, line := range strings.Split(resultSheet, ",") {
		entries, err := splitMatchData(line)
		if err != nil {
			return fmt.Sprintf("Error(%v):%s", err, line)
		}
		var game teamData
		if entries.team[0] == toFind {
			game, err = parseTeamData(entries, 0)
		} else if entries.team[1] == toFind {
			game, err = parseTeamData(entries, 1)
		} else {
			continue
		}
		if err != nil {
			return fmt.Sprintf("Error(%v):%s", err, line)
		}
		// The team to find was in the game. Add the result of the game.
		delta := game.pointsScored - game.pointsConceded
		switch {
		case delta > 0:
			numberMatchesWon++
			rank += 3
		case delta == 0:
			numberMatchesDrawn++
			rank++
		case delta < 0:
			numberMatchesLost++
		}
		totalPointsScored += game.pointsScored
		totalPointsConceded += game.pointsConceded
	}
	// Prepare the result string in the given format
	if numberMatchesWon == 0 && numberMatchesDrawn == 0 && numberMatchesLost == 0 {
		return toFind + ":This team didn't play!"
	} else {
		return fmt.Sprintf("%s:W=%d;D=%d;L=%d;Scored=%d;Conceded=%d;Points=%d", toFind, numberMatchesWon, numberMatchesDrawn, numberMatchesLost, totalPointsScored, totalPointsConceded, rank)
	}
}

File: 6_kyu/Ranking_NBA_teams.lua
local solution = {}

-- Collect results of NBA team "toFind"
-- Arguments:
--    resultSheet - A string containing list of all matches separated by ",".
--    toFind - Name of the team to find
-- Return:
--    Results of the team.
function solution.nbaCup(resultSheet, toFind)
  if type(resultSheet) ~= "string" or type(toFind) ~= "string" then
    -- invalid arguments
    return
  end
  if #resultSheet == 0 or #toFind == 0 then
    -- empty arguments
    return ""
  end
  -- terminate all matches by a ','
  resultSheet = resultSheet .. ','
  -- scan for decimal point number error
  for err in resultSheet:gmatch("[^.,]+%d+%.%d+[^.,]*,") do
    return "Error(float number):" .. err:sub(1,-2)
  end
  -- collect the results of the team "toFind" in the local variables
  local matches_won = 0
  local matches_draws = 0
  local matches_lost = 0
  local points_scored = 0
  local points_conceded = 0
  local points = 0
  for t1,p1,t2,p2 in resultSheet:gmatch("%s*([%s%w]+)%s+(%d+)%s+([%w%s]+)%s+(%d+)%s*,") do
    if toFind == t2 then
      t1,t2 = t2,t1
      p1,p2 = p2,p1
    end
    if toFind == t1 then
      -- parse only details if this is the team to find
      p1 = tonumber(p1)
      p2 = tonumber(p2)
      if p1 > p2 then
        matches_won = matches_won + 1
        points = points + 3
      elseif p2 > p1 then
        matches_lost = matches_lost + 1
      else
        matches_draws = matches_draws + 1
        points = points + 1
      end
      points_scored = points_scored + p1
      points_conceded = points_conceded + p2
    end
  end
  -- Format result string
  if matches_won+matches_draws+matches_lost == 0 then
    return toFind .. ":This team didn't play!"
  else
    return string.format("%s:W=%d;D=%d;L=%d;Scored=%d;Conceded=%d;Points=%d",toFind,matches_won,matches_draws,matches_lost,points_scored,points_conceded,points)
  end
end

return solution
